blockchain/types/src/avr_history.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Types for reading and writing historical Attestation Verification
+//! Reports from MobileCoin consensus enclaves
+
+use alloc::{string::String, vec::Vec};
+use hex::ToHex;
+use mc_attest_core::VerificationReport;
+use mc_crypto_keys::Ed25519Public;
+use mc_common::HashMap;
+use toml::{map::Map, value::Value};
+use url::Url;
+
+/// Object representing AVR history for all consensus nodes
+pub struct AVRHistoricalRecords {
+    /// AVR Records
+    pub avr_records: HashMap<Url, Vec<AVRHistoricalRecord>>,
+}
+
+impl AVRHistoricalRecords {
+    /// Create new records object
+    pub fn new(avr_records: HashMap<Url, Vec<AVRHistoricalRecord>>) -> Self {
+        Self { avr_records }
+    }
+
+    /// Get a toml representation of the AVR historical records
+    pub fn export_toml_string(&self) -> Result<String, toml::ser::Error>{
+        toml::to_string(&Map::from(self))
+    }
+}
+
+/// Stores a single historical AVR for each tuple of responder_id
+/// & block signing key  in the format specified in draft MCIP #43
+pub struct AVRHistoricalRecord {
+    /// Uri of the consensus node
+    pub node_url: Url,
+
+    /// Block signing key of the consensus enclave
+    pub public_key: Option<Ed25519Public>,
+
+    /// Block the AVR for the signing key becomes valid
+    pub start_block: u64,
+
+    /// Final block the AVR for the signing key is valid
+    pub end_block: u64,
+
+    /// Hex of the AVR bytes
+    pub avr: Option<VerificationReport>,
+}
+
+impl AVRHistoricalRecord {
+    /// Create new AVRHistoricalRecord
+    pub fn new(
+        node_url: &Url,
+        public_key: &Option<Ed25519Public>,
+        start_block: u64,
+        end_block: u64,
+        avr: Option<VerificationReport>,
+    ) -> Self {
+        Self {
+            node_url: node_url.clone(),
+            public_key: *public_key,
+            start_block,
+            end_block,
+            avr,
+        }
+    }
+}
+
+impl Default for AVRHistoricalRecords {
+    fn default() -> Self {
+        let avr_records: HashMap<Url, Vec<AVRHistoricalRecord>> = HashMap::default();
+        Self::new(avr_records)
+    }
+}
+
+// Rust <--> TOML format mapping for all historical AVRs
+impl From<&AVRHistoricalRecords> for Map<String, Value> {
+    fn from(src: &AVRHistoricalRecords) -> Self {
+        let mut toml = Map::new();
+        for (url, records) in src.avr_records.iter() {
+            let maps: Vec<Value> = records.iter().map(|x| Value::Table(Map::from(x))).collect();
+            toml.insert(url.as_str().into(), Value::Array(maps));
+        }
+        toml
+    }
+}
+
+// Rust <--> TOML format mapping for a single AVR
+impl From<&AVRHistoricalRecord> for Map<String, Value> {
```

Comment:
If we rename the Rust fields like I'm suggesting, we can replace this with `#[derive(Serialize, Deserialize)` on `AvrRecord`/`AVRHistoricalRecord`

Reply:
Agreed, there shouldn't be a need to manually implement toml serialization/deserialization. 
For stuff like encoding AVRs, you can see https://github.com/mobilecoinfoundation/mobilecoin/blob/master/consensus/service/config/src/tokens.rs#L128 as an example of how to override the serialization/deserialization behavior of a specific field.

Reply:
Sure, I'm on board, thanks for the information.